frontend/typed_args.py

```diff
@@ -113,32 +114,124 @@ def RestWords(self):
 
     ### Typed positional args
 
-    # TODO: may need location info
     def PosStr(self):
         # type: () -> str
-        return None  # TODO
+        try:
```

Comment:
Hm I would like to see this API used in context.  We need to test it out before knowing if it's right

And particularly have a demo of the error message it produces e.g.


- `len()`  --  expected 1 argument, got 0
- len(1, 2)` -- expected 1 argument, got 2
- `_match('group')` and `_match(1)` -- I think that's our API for regexes, it takes either a string or int, like Python's `m.group(1)` and `m.group('name')`

"No arguments left" isn't a good error message

- I think it should use a counter internally, and also avoid IndexError
-  I think we can inline `MustBe*` into those functions, or at least put them in this file, and eliminate other usages

i.e. we should try to convert `args.Spec` and `MustBe*` and `DoesNotAccept` all into this single API

---

But if you want to do something more minimal, I would just start with `len()` and then see what the errors look like


i.e.  I try generalize from specific examples

Reply:
> "No arguments left" isn't a good error message
> - I think it should use a counter internally, and also avoid IndexError

Oh sure. This message could be better. I don't see how catching `IndexError` gets in the way of this or why we need extra state, though. What if we add a new method, like `ExpectNPos(n)` (and maybe `ExpextAtLeastN(n)`) and have the caller use that before any of the other `Pos*()`

> i.e. we should try to convert args.Spec and MustBe* and DoesNotAccept all into this single API

It won't be useful to have those helpers for use elsewhere? i.e., it will be awkward to have to pack an individual `value_t` into a list and construct a `Reader` in non-function-like/non-proc-like contexts. Or do you not think there will be many cases like that?

Reply:
Whoops. Missed this one.

> Hm I would like to see this API used in context. We need to test it out before knowing if it's right

The new unit tests don't capture this well enough? What else do you have in mind?

Reply:
The unit tests are helpful, but synthetic usages ... At least we want to use it in len() and a few other places, and see what the API looks like

And I'll push and look at the error messages

At the very least it should replace the typed_args.Spec usages, which we all agreed about, and we can delete that API

 ExpectNPos() is redundant, so I would try to avoid that, similar to typed_args.Spec

---


As far exceptions, it may not matter here, but we should avoid using them internally -- i.e. when not caught at the top level and exiting the program.

It's the only case where C++ is slower than Python, see here it's 1.7x slower, and other cases are 100x faster

https://www.oilshell.org/release/0.17.0/benchmarks.wwz/mycpp-examples/

(C++ exceptions are optimized for the non-throw path, so the throw path is very slow)

---

The imperative nature of the API should also be hidden in the Done() error messages

Reply:
Ah damn! I didn't realize c++ exceptions were so much slower 👀 Will make a few tweaks